Feature/member

minju412 opened the pull request:
### 변경 사항

- entity 변경 (erd 참고)
- jwt를 이용한 회원가입 및 로그인
- exception 추가
- cors 설정

minju412 committed:
fix: user entity 변경

minju412 committed:
feat: jwt 로그인
- username(=아이디)와 password로 로그인

minju412 committed:
feat: 사용자 정보 조회

minju412 committed:
fix: entity 수정

minju412 committed:
fix: entity 수정

minju412 committed:
feat: 회원 정보 수정

minju412 committed:
fix: security config 수정

minju412 committed:
fix: 회원가입 exception 추가

minju412 committed:
refactor: exception

minju412 committed:
rename: exception 폴더명 변경

minju412 committed:
fix: exception 추가

minju412 committed:
fix: response 필드 변경

minju412 committed:
fix: cors 설정

Jeyoung-Park commented:
@minju412 고생 많으셨니다!
pr 단위는 작게 가져가면 더 좋을 것 같습니다. 이 pr의 경우 39개의 파일이 변경된 큰 pr로 보이는데, 큰 pr은 리뷰하기도 쉽지 않고 관리하기도 좋지 않아서 지양하는 게 좋을 것 같아요 :)

Jeyoung-Park commented on src/main/java/naem/server/config/CustomAccessDeniedHandler.java:
```diff
@@ -0,0 +1,61 @@
+        Collection<GrantedAuthority> authorities = member.getAuthorities();
+
+        if (hasRole(authorities, MemberRole.ROLE_NOT_PERMITTED.name())) {
+            response.setMessage("인증되지 않은 사용자입니다.");
```
response 메시지 이외에 status code도 정의되어 있나요?

minju412 replied:
@Jeyoung-Park 토큰 저장방식 수정하면서 같이 변경하도록 하겠습니다.

Jeyoung-Park commented on src/main/java/naem/server/config/CustomAccessDeniedHandler.java:
```diff
@@ -0,0 +1,61 @@
+
+        if (hasRole(authorities, MemberRole.ROLE_NOT_PERMITTED.name())) {
+            response.setMessage("인증되지 않은 사용자입니다.");
+            // throw new CustomException(UNAUTHORIZED_USER);
```
@minju412 사용하지 않는 주석은 지우는 게 코드 관리하는 데에 좋을 것 같아요

Jeyoung-Park commented on src/main/java/naem/server/config/JwtRequestFilter.java:
```diff
@@ -0,0 +1,105 @@
+    private JwtUtil jwtUtil;
+
+    @Autowired
+    private CookieUtil cookieUtil;
```
@minju412 저희는 모바일 환경이라 cookie가 필요하지 않는 걸로 알고 있습니다. 클라이언트에서 쿠키로 보내줘야 하는 정보가 있나요?

minju412 replied:
@Jeyoung-Park 쿠키로 토큰 정보를 보내주는 방식을 사용하려고 했는데 모바일 환경에서의 쿠키 사용에 대해 더 알아보도록 하겠습니다!

Jeyoung-Park commented on src/main/java/naem/server/config/JwtRequestFilter.java:
```diff
@@ -0,0 +1,105 @@
+        String username = null;
+        String jwt = null;
+        String refreshJwt = null;
+        String refreshUname = null;
```
@minju412 refreshUname은 refreshUsername을 줄이신 건가요? refreshUsername으로 한다고 해도 변수명이 길지 않기 떄문에 refreshUsername으로 가면 어떨지 제안드립니다! 이건 중요도가 낮은 제안이라 굳이 반영 안 하셔도 됩니다 :)

minju412 replied:
@Jeyoung-Park 넵 맞습니다! 반영했습니다:)

Jeyoung-Park commented on src/main/java/naem/server/domain/BoardType.java:
```diff
@@ -0,0 +1,5 @@
+package naem.server.domain;
+
+public enum BoardType {
```
게시판을 enum 타입으로 선언한 건 좋습니다 👍

minju412 replied:
감사합니다ㅎㅎ

Jeyoung-Park commented on src/main/java/naem/server/domain/Comment.java:
```diff
@@ -26,8 +35,16 @@ public class Comment {
+
+    @Temporal(TemporalType.TIMESTAMP)
+    @UpdateTimestamp
+    private Date updateAt;
```
삭제와 관련된 delete, deletedAt 같은 필드는 없어도 괜찮은가요? 추가로 댓글에 있는 좋아요는 기능이 어떻게 논의되었나요? 아직 논의되지 않았다면 슬랙에 글 올려주시면 좋을 것 같습니다 @minju412

minju412 replied:
@Jeyoung-Park 댓글에도 삭제 관련 필드가 있는게 좋겠네요! 추가했습니다🙂 
좋아요 기능에 대해서도 리마인드 감사합니다! 바로 글 올리도록 하겠습니다.

Jeyoung-Park commented on src/main/java/naem/server/domain/member/Member.java:
```diff
@@ -10,50 +10,71 @@
-    private String name; // 사용자 이름
+    @Column(unique = true)
+    @Size(min = 3, max = 50)
+    private String username;
```
@minju412 username은 사용자 이름인가요? 사용자 이름은 unique하지 않을 수도 있습니다.

minju412 replied:
@Jeyoung-Park username은 로그인시 사용하는 아이디입니다!

Jeyoung-Park commented on src/main/java/naem/server/domain/member/Member.java:
```diff
@@ -10,50 +10,71 @@
 
+    // @JsonProperty(access = JsonProperty.Access.WRITE_ONLY)
     @NotBlank
+    @Size(min = 3, max = 100)
```
이 size는 text의 길이를 의미하는 건가요? 만약 그렇다면 비밀번호 유효성을 참고하셔서 이 부분을 다시 정의하셔야할 듯 합니다! @minju412

minju412 replied:
@Jeyoung-Park 앗 네네! 생각해보니 유효성 부분을 잊었네요ㅜ 참고하여 변경하도록 하겠습니다!

Jeyoung-Park commented on src/main/java/naem/server/domain/member/SecurityMember.java:
```diff
@@ -0,0 +1,14 @@
+import org.springframework.security.core.authority.AuthorityUtils;
+import org.springframework.security.core.userdetails.User;
+
+public class SecurityMember extends User {
```
@minju412 이 SecurityMember 클래스는 어떤 역할을 하는지 설명 요청드려도 괜찮을까요?

minju412 replied:
@Jeyoung-Park authorities를 가지고 있는 인증된 사용자를 의미하는데, User라는 spring security에 정의되어 있는 클래스를 상속받아 구현하였습니다.

Jeyoung-Park commented on src/main/java/naem/server/domain/member/Member.java:
```diff
@@ -63,8 +84,9 @@ public class User {
+    private Boolean isDeleted;
+
+    @Temporal(TemporalType.TIMESTAMP)
+    private Date deleteAt;
```
@minju412 생각해보니 장애 유형 필드도 존재해야 합니다!

Jeyoung-Park replied:
유저 프로필 사진 필드도 없는 것 같네요

minju412 replied:
@Jeyoung-Park 장애 유형 필드를 놓쳤네요! 확인 감사합니다:) 
이미지는 용량이 커서 db에 직접 삽입하지 않는다고하여, 파일은 s3에 저장하고 file path를 db에 저장하는 방식으로 진행하려고 합니다! (filePath는 Member 클래스에 정의되어 있습니다.)

Jeyoung-Park commented on src/main/java/naem/server/domain/member/dto/PatchMemberDto.java:
```diff
@@ -0,0 +1,12 @@
+
+@Getter
+@Setter
+public class PatchMemberDto {
```
@minju412 

patch할 떄 추가적으로 유저 프로필 사진과 관심 키워드(장애유형)도 수정이 가능해야 합니다.

![image](https://user-images.githubusercontent.com/59302192/183416641-2aaa0270-6a0d-4b60-b2c1-ba0d75ff1691.png)

minju412 replied:
@Jeyoung-Park 관심 키워드와 장애 유형이 같은 필드인거 맞을까요? 그렇다면 member와 post에 사용되고 있는 장애유형 필드를 관심 키워드 필드로 네이밍을 바꾸는게 나을까요..?!

Jeyoung-Park replied:
@minju412 음 같은 필드로 보는 게 맞을 것 같습니다. 네이밍 바꾸는 게 공수가 크지 않다면 바꾸면 좋을 것 같습니다!

minju412 replied:
@Jeyoung-Park 

<img width="343" alt="스크린샷 2022-08-09 오후 2 04 19" src="https://user-images.githubusercontent.com/59405576/183568491-b7a3b7ab-ba9a-482c-a1aa-f118ea4d941e.png">
tag가 post에도 사용되고있기 때문에 이름을 바꾸면 같이 바꿔야하는게 많아서, tag의 네이밍을 그대로 가져가고
위 사진처럼 member의 관심키워드를 member_tag라고 네이밍하는건 어떨까요..?!

Jeyoung-Park replied:
@minju412 넵 좋습니다!

Jeyoung-Park commented on src/main/java/naem/server/exception/ErrorCode.java:
```diff
@@ -0,0 +1,33 @@
+
+@Getter
+@AllArgsConstructor
+public enum ErrorCode {
```
@minju412 에러 상태들 이렇게 정의하는 건 좋은 것 같습니다 👍

minju412 replied:
@Jeyoung-Park 넵 감사합니다!😃

Jeyoung-Park commented on src/main/java/naem/server/repository/MemberRepository.java:
```diff
@@ -0,0 +1,19 @@
+
+public interface MemberRepository extends CrudRepository<Member, Long> {
+
+    Member findByUsername(String username);
```
@minju412 아래 메서드들 모두 하나의 값만을 리턴하므로 findBy~ -> findOneBy로 네이밍하면 어떨지 제안드립니다!

minju412 replied:
@Jeyoung-Park 해당 메서드는 제가 직접 정의한게 아니라 필드를 조회하기 위해서 " findBy + Entity 필드이름 " 규칙에 맞는 메서드를 추가한 것이기 때문에 변경하기 어렵습니다🥲

Jeyoung-Park commented on src/main/java/naem/server/service/AuthServiceImpl.java:
```diff
@@ -1,65 +1,81 @@
 
 @Service
 @Slf4j
 public class AuthServiceImpl implements AuthService {
```
@minju412 AuthServiceImpl 클래스 네이밍 관련한 질문입니다. 뒤에 Impl은 무엇을 의미하나요?

minju412 replied:
@Jeyoung-Park AuthService 인터페이스의 구현체(Implementation)을 의미합니다.

Jeyoung-Park commented on src/main/java/naem/server/service/AuthServiceImpl.java:
```diff
@@ -1,65 +1,81 @@
-        user.setPassword(saltUtil.encodePassword(salt, password));
-        userRepository.save(user);
+        member.setSalt(new Salt(salt));
+        member.setPassword(saltUtil.encodePassword(salt, password));
```
@minju412 암호화하는 것 좋습니다! 이 암호화방식은 암호화를 했을 떄 복호화도 할 수 있나요?

minju412 replied:
@Jeyoung-Park 아니요! 복호화는 할 수 없습니다. 복호화가 가능한 암호화는 보안에 굉장히 취약한 방식입니다..!
비밀번호 분실 시 새로운 비밀번호로 변경할 수 있도록 진행할 예정입니다.

Jeyoung-Park replied:
@minju412 네넵! 맞습니다 저도 그 부분에 대한 걱정 떄문에 여쭤봤었는데 잘 해주셨군요..!! 👍

Jeyoung-Park commented on src/main/java/naem/server/service/AuthServiceImpl.java:
```diff
@@ -1,65 +1,81 @@
+        result = Optional.ofNullable(memberRepository.findByNickname(member.getNickname()));
+        result.ifPresent(m -> {
+            throw new CustomException(CONFLICT_NICKNAME);
+        });
```
@minju412 중복 검사 이외에 유효성 검증은 없나요?

minju412 replied:
@Jeyoung-Park 유효성 검증이 빠져있어 추가하도록 하겠습니다!

Jeyoung-Park replied:
@minju412 넵 유효성 검사 부분은 이번 pr에서 적용하기 양이 너무 많다면 다음 pr 때 적용 부탁드려요

Jeyoung-Park commented on src/main/java/naem/server/service/MyUserDetailsService.java:
```diff
@@ -0,0 +1,31 @@
+    @Override
+    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
+
+        System.out.println("MyUserDetailsService.loadUserByUsername");
```
@minju412 이런 로그들도 지워주는 게 깔끔한 코드 유지하는 데에 좋습니다!

minju412 replied:
@Jeyoung-Park 넵 제거했습니다!:)

Jeyoung-Park replied:
@minju412 제거된 내용은 어디에 반영해주셨나요?

minju412 replied:
@Jeyoung-Park 로컬에 반영되어있고 곧 푸시할 예정입니다!

Jeyoung-Park replied:
@minju412 넵 푸시되면 멘션 한 번 주세요!

Jeyoung-Park commented on src/main/java/naem/server/service/util/JwtUtil.java:
```diff
@@ -0,0 +1,81 @@
+public class JwtUtil {
+
+    public final static long TOKEN_VALIDATION_SECOND = 1000L * 60 * 30; // 30분
+    public final static long REFRESH_TOKEN_VALIDATION_SECOND = 1000L * 60 * 60 * 24 * 7; // 7일
```
@minju412 로그인 이후 얼마 동안 유지될지를 이 값으로 결정하는 거죠? 이 것도 슬랙 채널에 올려주셔서 다른 분들 의견 구하면 좋을 것 같아요!

minju412 replied:
@Jeyoung-Park 그게 좋겠네요! 의견 올리도록 하겠습니다.

Jeyoung-Park commented on src/test/java/naem/server/AuthServiceTest.java:
```diff
@@ -1,11 +1,14 @@
 package naem.server;
```
테스트 코드 작성하시는 거 좋네요!

minju412 replied:
@Jeyoung-Park 감사합니다!🙂

Jeyoung-Park commented:
@minju412 추가로 스웨거 문서는 같이 작성 안 하시나요?

minju412 committed:
fix: 코드리뷰 반영
- 회원가입 시 유효성 검사 및 exception 추가
- Member의 username, nickname 필드에 길이 제한 추가
- Member에 관심 키워드 필드(member_tag) 추가
- PatchMemberDto에 filePath(프로필 이미지) 필드 추가
- Comment에 삭제 관련 필드(isDeleted, deleteAt) 추가
- 불필요한 주석 제거

minju412 committed:
fix: 토큰 저장 방식 변경
- 기존: 쿠키 사용
- 변경: 헤더에 추가

minju412 committed:
fix: exception 추가

minju412 commented:
> @minju412 고생 많으셨니다! pr 단위는 작게 가져가면 더 좋을 것 같습니다. 이 pr의 경우 39개의 파일이 변경된 큰 pr로 보이는데, 큰 pr은 리뷰하기도 쉽지 않고 관리하기도 좋지 않아서 지양하는 게 좋을 것 같아요 :)

넵 알겠습니다!:D

minju412 commented:
> @minju412 추가로 스웨거 문서는 같이 작성 안 하시나요?

@Jeyoung-Park 스웨거 문서를 자동으로 생성되게끔 하였는데
http://13.124.66.164:8080/swagger-ui/ 
혹시 이거 말고 별도로 설정이 필요한 부분이 있을까요..?!

Jeyoung-Park commented:
> > @minju412 추가로 스웨거 문서는 같이 작성 안 하시나요?
> 
> @Jeyoung-Park 스웨거 문서를 자동으로 생성되게끔 하였는데
> http://13.124.66.164:8080/swagger-ui/ 
> 혹시 이거 말고 별도로 설정이 필요한 부분이 있을까요..?!

아 그렇군요 예전에 노드 했을 때 yaml 파일로 작성해줬었던 같아 여쭤봣습니다!

minju412 committed:
"feat: 내 정보 조회 api 추가"
- 내가 쓴 글과 댓글은 게시글 CRUD, 댓글 CRUD 구현 후 추가 예정

minju412 commented:
> 



> 



> 



> > > @minju412 추가로 스웨거 문서는 같이 작성 안 하시나요?
> > 
> > 
> > @Jeyoung-Park 스웨거 문서를 자동으로 생성되게끔 하였는데
> > http://13.124.66.164:8080/swagger-ui/
> > 혹시 이거 말고 별도로 설정이 필요한 부분이 있을까요..?!
> 
> 아 그렇군요 예전에 노드 했을 때 yaml 파일로 작성해줬었던 같아 여쭤봣습니다!

yaml 파일로 작성하는 법도 있군요! 관련해서 알아보도록 하겠습니다:)

Jeyoung-Park commented:
> > 
> 
> > 
> 
> > 
> 
> > > > @minju412 추가로 스웨거 문서는 같이 작성 안 하시나요?
> > > 
> > > 
> > > @Jeyoung-Park 스웨거 문서를 자동으로 생성되게끔 하였는데
> > > http://13.124.66.164:8080/swagger-ui/
> > > 혹시 이거 말고 별도로 설정이 필요한 부분이 있을까요..?!
> > 
> > 
> > 아 그렇군요 예전에 노드 했을 때 yaml 파일로 작성해줬었던 같아 여쭤봣습니다!
> 
> yaml 파일로 작성하는 법도 있군요! 관련해서 알아보도록 하겠습니다:)

앗 넵 자동으로 업데이트되는 부분이라면 신경 안 써주셔도 됩니다!

Jeyoung-Park commented on src/main/java/naem/server/config/JwtAuthenticationEntryPoint.java:
```diff
@@ -0,0 +1,27 @@
+        HttpServletResponse response,
+        AuthenticationException ex
+    ) throws IOException {
+        // System.out.println(request.getRequestURI());
```
@minju412 이런 데도 주석 지워지면 좋을 것 같습니다! 나중에 사용할 수도 있는 부분이라면 남겨 주세요!

minju412 replied:
@Jeyoung-Park 주석 제거했습니다!

Jeyoung-Park commented on src/main/java/naem/server/config/JwtTokenProvider.java:
```diff
@@ -0,0 +1,111 @@
+    }
+
+    private UsernamePasswordAuthenticationToken getUsernamePasswordAuthenticationToken(String token) {
+        String userPrincipal = Jwts.parser().setSigningKey(SECRET_KEY).parseClaimsJws(token).getBody().getSubject();
```
@minju412 parseClaimJws -> parseClaimJwt 오타 아닌지 확인 부탁드려요!

minju412 replied:
@Jeyoung-Park jws는 jwt의 하위 개념으로 parseClaimJws가 맞습니다!:D

Jeyoung-Park commented on src/main/java/naem/server/domain/member/dto/CoreRes.java:
```diff
@@ -0,0 +1,12 @@
+
+@Getter
+@AllArgsConstructor
+public class CoreRes {
```
@minju412 이 클래스 repsonse 공통 객체인가요?

minju412 replied:
@Jeyoung-Park 넵 맞습니다! 그런데 현재 사용하고 있지 않은 객체라 삭제했습니다!

Jeyoung-Park commented on src/main/java/naem/server/domain/member/dto/SignUpReq.java:
```diff
@@ -0,0 +1,70 @@
+@Setter
+@ToString
+@NoArgsConstructor
+public class SignUpReq {
```
@minju412 이번에 유효성까지 챙겨주셨군요! 최고입니다 👍 👍 👍

minju412 replied:
@Jeyoung-Park 넵 감사합니다!ㅎㅎ🙇‍♀️
다만 패스워드에서 16자 이하로 설정해야하는데, db에는 암호화 되어 긴 문자열이 들어가기때문에 16자 이하로 설정할 수가 없더라고요..ㅜㅜ 
이 부분은 일단은 넘어가고 mvp 이후에 다시 생각해도 될지 여쭤보고 싶습니다..!

Jeyoung-Park replied:
@minju412 넵 알겠습니다. 생각해보니 클라에서 데이터 보낼 때 한 번 암호화해서 보내려고 생각 중이어서 비밀번호 관련된 유효성 로직은 뺴주셔도 좋을 것 같아요!

minju412 replied:
@Jeyoung-Park 넵 알겠습니다!

minju412 replied:
@Jeyoung-Park 혹시 비밀번호 유효성에서 길이 제한 말고 아예 전부를 빼면 될까요..?!

Jeyoung-Park commented on src/main/java/naem/server/service/AuthServiceImpl.java:
```diff
@@ -2,80 +2,138 @@
+    private final JwtTokenProvider jwtTokenProvider;
+    private final AuthenticationManager authenticationManager;
+    private final RedisTemplate<String, String> redisTemplate;
+    private long REFRESH_TOKEN_EXPIRE_TIME = 1000L * 60 * 60 * 24 * 7; // 7일
```
@minju412 refreshtoken expire time은 위에서도 선언되었던 것 같은데 import해서 사용할 수는 없나요?

minju412 replied:
@Jeyoung-Park 
getter를 이용해 AuthServiceImpl에서 JwtTokenProvider에 정의된 REFRESH_TOKEN_EXPIRE_TIME을 사용하려고 시도했으나, JwtTokenProvider의 REFRESH_TOKEN_EXPIRE_TIME이 초기화가 안되어있을 수 있다고 빌드가 실패하네요ㅠ
우선은 그대로 진행하도록 하겠습니다..!🥲

Jeyoung-Park commented on src/main/java/naem/server/web/MemberController.java:
```diff
@@ -34,56 +26,46 @@
+    // 내 정보 조회
+    @GetMapping("/profile")
+    public ProfileRes profile(@AuthenticationPrincipal UserDetails userDetails) throws UserNotFoundException {
+        System.out.println("userDetails = " + userDetails);
```
@minju412 여기도 불필요한 프린트문은 지우는 게 좋지 않을까 싶습니다!

minju412 replied:
@Jeyoung-Park 앗 잊고 있었네요..ㅎㅎ 제거했습니다!

Jeyoung-Park reviewed (approved):
고생 많으셨습니다 🙏 🙏 🙏

minju412 committed:
fix: 코드리뷰 반영
- 불필요한 주석, 로그 제거
- 불필요한 객체 삭제

minju412 committed:
Remove: 불필요한 api 제거

minju412 committed:
fix: ProfileDto 변경

minju412 commented:
> 고생 많으셨습니다 🙏 🙏 🙏

감사합니다!!!🙇‍♀️🙇‍♀️
